src/java/com/puppetlabs/jruby_utils/pool/ReferencePool.java

```diff
@@ -0,0 +1,526 @@
+    // currently exist in the queue.
+    private final Condition borrowsAvailable = borrowLock.newCondition();
+
+    // Condition signaled when the pool has been unlocked.  Awaited when a
+    // request has been made to borrow an reference or lock the pool but the pool
+    // is currently locked.
+    private final Condition poolNotLocked = borrowLock.newCondition();
+
+    // Condition signaled when all borrowed references have been returned
+    // to the pool. Awaited when we are preparing to delete the instance
+    // and need to make sure it is not in use.
+    private final Condition instanceNotBorrowed = borrowLock.newCondition();
+
+    // The JRuby instance that this pool hands out references to
+    private volatile E instance;
+
+    // How many times the JRuby instance can be borrowed at once
+    private int maxBorrowCount;
+
+    // Current number of references to the pool held out in the world.
+    // Updates to this need to be visible to all threads.
+    private volatile AtomicInteger currentBorrowCount;
+
+    // Thread which currently holds the pool lock.  null indicates that
+    // there is no current pool lock holder.  Using the current Thread
+    // object for tracking the pool lock owner is comparable to what the JDK's
+    // `ReentrantLock` class does via the `AbstractOwnableSynchronizer` class:
+    //
+    // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/locks/ReentrantLock.java#l164
+    // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/locks/AbstractOwnableSynchronizer.java#l64
+    //
+    // Unlike the `AbstractOwnableSynchronizer` class implementation, we marked
+    // this variable as `volatile` because we couldn't convince ourselves
+    // that it would be safe to update this variable from different threads and
+    // not be susceptible to per-thread / per-CPU caching causing the wrong
+    // value to be seen by a thread.  `volatile` seems safer and doesn't appear
+    // to impose any noticeable performance degradation.
+    private volatile Thread poolLockThread = null;
+
+    // Holds a poison pill object for errors and shutdowns
+    // If not null, takes priority over any pool instance when a call to
+    // borrowItem is made. Returns made using releaseItem are ignored if the
+    // released item is the poison pill already stored here
+    private volatile E pill;
+
+    /**
+     * Create a "pool" of handles to a Jruby instance.
+     *
+     * @param maxBorrows the max number of instance refs that can be handed out
+     */
+    public ReferencePool(int maxBorrows) {
+        this.instance = null;
+        this.maxBorrowCount = maxBorrows;
+        this.currentBorrowCount = new AtomicInteger(0);
+    }
+
+    @Override
+    public void register(E e) {
+        final ReentrantLock lock = this.borrowLock;
+        lock.lock();
+        try {
+            if (instance != null) {
```

Comment:
I would find it easier to read if this was `this.instance` instead of just `instance` - it makes it clearer to me where `instance` is coming from. thought I definitely won't block on this - I don't know what conventions have been established in this repo already.

Reply:
So there are a _lot_ of places in this file that use fields without prefixing `this`. In my IDE, the syntax highlighting makes fields a different color than the local variables, but I notice that my color scheme in vim does not (and neither does GitHub). I don't have a strong preference here, but many of the other places will get significantly more verbose if we consistently add `this` everywhere.

I'd be interested to hear if others have a policy for approaching this beyond "any time there's potential shadowing" (most common in constructors in my experience). It's unclear to me what heuristic the original Java authors in this repo were following.